klippy/extras/ads1220.py

```diff
@@ -0,0 +1,203 @@
+# ADS1220 Support
+#
+# Copyright (C) 2024 Gareth Farrington <gareth@waves.ky>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+import collections
+import logging, struct
+from . import bulk_sensor, bus
+from .bulk_adc_sensor import (BulkSensorAdc, TimestampHelper)
+
+#
+# Constants
+#
+BYTES_PER_SAMPLE = 4  # samples are 4 byte wide unsigned integers
+MAX_SAMPLES_PER_MESSAGE = bulk_sensor.MAX_BULK_MSG_SIZE // BYTES_PER_SAMPLE
+UPDATE_INTERVAL = 0.10
+RESET_CMD = 0x06
+START_SYNC_CMD = 0x08
+RREG_CMD = 0x20
+WREG_CMD = 0x40
+NOOP_CMD = 0x0
+RESET_STATE = bytearray([0x0, 0x0, 0x0, 0x0])
+
+
+# turn bytearrays into pretty hex strings: [0xff, 0x1]
+def hexify(byte_array):
+    return "[%s]" % (", ".join([hex(b) for b in byte_array]))
+
+
+class ADS1220(BulkSensorAdc):
```

Comment:
I'd generally prefer not to define base classes like this in python - in particular when the base class is in another python file.  In Klipper the coding convention is to not touch the internal member variables of a class if that class is defined in another python file.  Alas, in python, when using base classes like this, it becomes really difficult to identify and maintain these types of dependencies between class instances.

I understand that in this case, the base class has no member variables and does not access any member variables.  However, it also seems like the base class just serves as documentation, and I think it would be easier to maintain the code going forward if comments were used for that purpose instead of a base class.

Reply:
Well I had to try 🤷‍♂️ 

I'd call this an interface, python calls this an [Abstract Base Class](https://docs.python.org/2.7/glossary.html#term-abstract-base-class) but cant explain what they mean without using the word interface. This is a mechanism. Its discoverable by your IDE (cmd+click takes you right there), it can be checked at build time and run time. With Python 3 type annotations you could enforce that methods accept this type. With 2.7 you can still do type checking at runtime.

I think you are missing out on a simple bit of tooling that could help clearly communicate your intentions in a discoverable way. Responding to everyone that didn't find the right code/doc/comment is exhausting in practice.

Reply:
Removed